File: amiXml2Array.py
import cookielib
import urllib
import urllib2
import xml.etree.ElementTree as elements
import string

class AmiXml2Array:
	def __init__(self,globalUrl,loginUrl):
		self.globalUrl = globalUrl
		self.loginUrl  = loginUrl 
    
	
	def login_ami_http(self,loginUrl):

		cj = cookielib.CookieJar()
		opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(cj))
		urllib2.install_opener(opener)
  		con = urllib2.Request(loginUrl)
		response = urllib2.urlopen(con)
		return response.read()

	def logoff_ami_http(self,globalUrl):
		con = urllib2.Request(self.globalUrl+'action=logoff')
		response = urllib2.urlopen(con)
		return response.read()


	def get_simple_action(self,action):
		
		self.login_ami_http(self.loginUrl)
		url = self.globalUrl+'action='+action;			
		con = urllib2.Request(url)
		response = urllib2.urlopen(con)			
		self.logoff_ami_http(self.globalUrl)			
		xml_nodes = response.read()
		json = self.parse_xml(xml_nodes)
		return json

	def get_queue_status_by_name(self, queueName):
		self.login_ami_http(self.loginUrl)
		url = self.globalUrl+'action=QueueStatus&queue='+queueName
		con = urllib2.Request(url)
		response = urllib2.urlopen(con)			
		self.logoff_ami_http(self.globalUrl)			
		xml_nodes = response.read()
		json = self.parse_xml(xml_nodes)
		return json

	def get_sipstatus_by_sip_peer(self,sipPeeer):
		self.login_ami_http(self.loginUrl)
		url = self.globalUrl+'action=SIPshowpeer&peer='+str(sipPeeer)
		con = urllib2.Request(url)
		response = urllib2.urlopen(con)			
		self.logoff_ami_http(self.globalUrl)			
		xml_nodes = response.read()
		json = self.parse_xml(xml_nodes)
		return json		

	def parse_xml(self,xmlString):
		newsitems = []
		itens = elements.fromstring(xmlString)
		# root = itens.getroot()

		for item in itens.findall('./response'):
		 	for child in item:
		 		newChild = string.replace(str(child.attrib),'"','')
		 		newsitems.append(newChild)
		
		outPutFormat = string.replace(str(newsitems),'"{','{')
		outPutFormat = string.replace(outPutFormat,'}"','}')
		
		return  string.replace(outPutFormat,'\'','"')


globalUrl = 'http://192.168.1.10:8088/mxml?'
loginUrl = 'http://192.168.1.10:8088/mxml?action=login&username=asterisk&secret=asterisk'
ami = AmiXml2Array(globalUrl,loginUrl)

#Exemplo de utilizacao
print(ami.get_simple_action('QueueStatus'))
print(ami.get_queue_status_by_name('fila_teste'))
print(ami.get_sipstatus_by_sip_peer('11004'))

File: amiXml2Array.php
<?php

/**
 * Simple classe de conexao e comandos(actions) ami utilizando php
 * Para habilitar , em manager.conf
 * enabled = yes 
 * webenabled = yes
 * Observar a config permit 
 * Ex liberar acesso para o ip 192.168.1.10
 * permit = 192.168.1.10/255.255.255.0
 * usuario de exemplo em manager.conf
 * [teste_asterisk]
 * secret = asterisk
 * read = call,dialplan
 * write = system,all
 * 
 * https://wiki.asterisk.org/wiki/display/AST/Allow+Manager+Access+via+HTTP
 */
class AmiXml2Array
{
    
    public $curl,$pathTmpFile,$loginUrl,$globalUrl,$options;
    
    /**
    * @access protected
    * Configuracoes das opcoes do curl
    */
    public function __construct()
    {
        
        $this->curl         = $this->curlStart();
        $tempFile           = tmpfile();
        $this->pathTmpFile  = stream_get_meta_data($tempFile)['uri'];
        $this->loginUrl     = '192.168.1.10:8088/mxml?action=login&username=teste_asterisk&secret=asterisk';
        $this->globalUrl    = '192.168.1.10:8088/mxml?';
        $this->options      =  [            
                                CURLOPT_COOKIESESSION   => TRUE,
                                CURLOPT_RETURNTRANSFER  => TRUE,
                                CURLOPT_TIMEOUT         => 30,
                                CURLOPT_HTTP_VERSION    => CURL_HTTP_VERSION_1_1,
                                CURLOPT_CUSTOMREQUEST   => "GET",
                                CURLOPT_COOKIEFILE      => $this->pathTmpFile
                            ];

    }
    
    public  function curlStart(){
        return curl_init();
    }

    /**    
     * Faz login no manager utilizando  curl 
    */
    public function loginAmiHttp()
    {
        
        $this->options[CURLOPT_URL]   = $this->loginUrl;
        
        $curl           = $this->curl;
        curl_setopt_array($curl, $this->options);
        $pathTmpFile    = $this->pathTmpFile;
        $data           = curl_exec($curl);

        $xml            = simplexml_load_string($data);
        $nodes          = [];
        foreach ($xml as $node) {
            $nodes[] = current($node->generic->attributes());
        }
        return compact('curl','pathTmpFile');        
    }    

    /**
     * Faz logoff do manager 
    */ 
    public function logoffAmiHttp()
    {
        $curl                               = $this->curl;
        $this->options[CURLOPT_URL]         = $this->globalUrl . 'action=logoff';
        $this->options[CURLOPT_COOKIEFILE]  = '';
        curl_setopt_array($this->curl, $this->options);
        $data = curl_exec($this->curl);
        return $data;
    }
     
    /**
     * Retorna um array com informacoes referentes a action consultada
     * @param string $action - Ver em  https://wiki.asterisk.org/wiki/display/AST/Asterisk+14+AMI+Actions
    */
    public function getSimpleAction($action,$object = false){

        $curlData       = $this->loginAmiHttp();
        if(!$curlData){
            return false;
        }

        $curl           = $curlData['curl'];
        $tmpFile        = $curlData['pathTmpFile'];
        $url            = $this->globalUrl . "action={$action}";        
        $this->options[CURLOPT_URL]   = $url;
        curl_setopt_array($curl, $this->options);
        $data   = curl_exec($curl);

        $this->logoffAmiHttp();
        $xml    = simplexml_load_string($data);
        $nodes = [];
        foreach ($xml as $node) {
            $nodes[] = current($node->generic->attributes());
        }
        
        if($object){
            return  json_encode($nodes);
        }
        return $nodes;
    }

    /**
     * Retorna um array com informacoes sobre a fila
     * @param string $queuName - O nome da fila
    */
    public function getQueueStatusByName($queueName,$object = false){

        $curlData       = $this->loginAmiHttp();
        if(!$curlData){
            return false;
        }

        $curl           = $curlData['curl'];
        $tmpFile        = $curlData['pathTmpFile'];
        $url            = $this->globalUrl . "action=QueueStatus&queue={$queueName}";        
        
        $this->options[CURLOPT_URL]   = $url;
        curl_setopt_array($curl, $this->options);
        $data   = curl_exec($curl);

        $this->logoffAmiHttp();
        $xml    = simplexml_load_string($data);
        $nodes = [];
        foreach ($xml as $node) {
            $nodes[] = current($node->generic->attributes());
        }

        if($object){
            return  json_encode($nodes);
        }
        return $nodes;
    }

    /**
     * Retorna um array com informacoes sobre a fila
     * @param string $sipPeeer - O peer a ser consultado(ex:1004)
    */
    public function getSipStatusBySipPeer($sipPeeer,$object = false){

        $curlData       = $this->loginAmiHttp();
        if(!$curlData){
            return false;
        }

        $curl           = $curlData['curl'];
        $tmpFile        = $curlData['pathTmpFile'];
        $url            = $this->globalUrl . "action=SIPshowpeer&peer={$sipPeeer}";        
        
        $this->options[CURLOPT_URL]   = $url;
        curl_setopt_array($curl, $this->options);
        $data   = curl_exec($curl);

        $this->logoffAmiHttp();
        $xml    = simplexml_load_string($data);
        $nodes = [];
        foreach ($xml as $node) {
            $nodes[] = current($node->generic->attributes());
        }
        
        if($object){
            return  json_encode($nodes);
        }
        return $nodes;
    }
}

/**
* More actions
* https://wiki.asterisk.org/wiki/display/AST/Asterisk+14+AMI+Actions
* Examples
*/

$ami = new AmiXml2Array();

//get status from all sip peers
//json
print_r($ami->getSimpleAction('SIPpeers',true));
//array
print_r($ami->getSimpleAction('SIPpeers'));

//queue status from all queues
//json
 print_r($ami->getSimpleAction('QueueStatus',true));
//array
print_r($ami->getSimpleAction('QueueStatus'));

// // //queue status from callcenter_1_fila_asdfasd_10
//json
print_r($ami->getQueueStatusByName('callcenter_1_fila_asdfasd_10',true));
//array
print_r($ami->getQueueStatusByName('callcenter_1_fila_asdfasd_10'));

//get status from peer 11004
//json
print_r($ami->getSipStatusBySipPeer('11004',true));
//array
print_r($ami->getSipStatusBySipPeer('11004'));
